### PR TITLE
Add "to accept" ConditionSet

### PR DESCRIPTION
**Pull request by rovermicrover**
## Feature Details
 I want to create a repeatable job based on @Hurleveur Unfettered Hai Duel mission framework. In order to do this the jobs will generally need to be offered, but not acceptable unless the player meets some preconditions.

The idea would be to create "Arena Duel" jobs in Pirate space that would require the player to have no escorts, and possible be in a certain class of ship. Thus the "to accept" conditions. It simply feeds into the existing Mission method CanAccept, just like "to offer" feeds into CanOffer. Wondering if this could help anyone currently, or if anyone has any feedback. Would have the exact same capabilities that "to offer" does.

## UI Screenshots
N/A

## Usage Examples
```
to accept
    "total ships" == 1
```

## Automated Tests Added
Trying to figure this out if its even possible. Not sure how jobs would work within the integrations test framework.

## Performance Impact
Just another ConditionSet check among many.

### CODE REVIEW

**Comment by Hurleveur** [comment]
I dont know if this is a good idea honestly because it could mean that a job is greyed out without any explanation, right? You could just fail the job upon accepting with an explanation message

**Comment by rovermicrover** [comment]
It would mimic the behavior of cargo of passenger jobs when you do not have room.

It is less of a surprise that way, as long as the requirements are laid out in the mission description.

**Comment by rovermicrover** [comment]
“to accept” effectively allows the CC to “fail early” with the player not even starting the mission.

My worry is with requiring a certain class of ship and then the player accepts but can’t avoid failing because said ship class isn’t in their fleet.

**Comment by warp-core** [comment]
Allowing content creators to do bad things is, I think, not a good enough reason not to add an otherwise good feature.

This does solve some problems I have seen in some places. A specific example that I can think of is the mission where we are given the Peregrine. The `to offer` of that mission requires we have the Remnant Capital License, but the mission has a `blocked` message explaining th need for that license. But, that's not how `blocked` messages work. A mission that can't be offered isn't `blocked`, a mission that can't be *accepted* on the other hand, is `blocked`.

It's possible, however, that the approach here is not the best one, but I can't see a better one, and I suspect this behaviour will not be easy for content creators to understand (the existing behaviour is already apparently unclear to some.)
I do not see any other immediate issues to raise relating to this, though.

**Comment by warp-core** [comment]
> We should have the blocked message display when you press on the accept button of the job board though? If that happens I'm down for this. Or some other way, maybe just add the blocked message to the description? idk but somehow show it

Or, just make sure people put the requirements in the description?
Blocked jobs already don't show a blocked message, they don't have one, the expectation is that the requirement is in the description, but it is already possible for a content creator to create a job and not include what the requirements are in the description.

**Comment by Hurleveur** [comment]
I don't see why both would be mutually exclusive? A mission description has to be very short btw and its a bit annoying if you always have to specify things even when the player does it already

**Comment by warp-core** [comment]
> I don't see why both would be mutually exclusive? A job description has to be very short btw and its a bit annoying if you always have to specify things even when the player does it already Also the jobs dont say anything because its obvious why its locked.

Perhaps missions with more complex requirements to accept are inappropriate for the job board anyway, then.
If you can't concisely convey why this job that is being offered can't be accepted, perhaps you shouldn't be offering it in the first place?

**Comment by Hurleveur** [comment]
My point is we should onnly display why its not accepted when its not accepted

**Comment by rovermicrover** [comment]
We could add a CC defined string called “failureExplanation” to ConditionSet, and then display that appended to the Job description when the job isn’t acceptable.

I think that satisfies @Hurleveur worry?

**Comment by Amazinite** [comment]
You might as well just include that information in the normal description, as Warp has said.

**Review by Hurleveur — APPROVED** [review]
Looks good code wise, you didn't forget anything that I can see. You could make an integration test for this but I'd say it would probably be a bit overkill? Adding a job that would offer on Ruin would just open up the job board where one could try to press accept but it's fine like this not like it can fail.
Apparently everyone agrees that the responsibility of it making sense should be on the CC so I'll fold in line x)

**Review by warp-core — APPROVED** [review]

**Comment by Amazinite** [comment]
If we want to get the behavior of showing a `blocked` dialog when a mission can offer but can't accept, then this line needs changed in PlayerInfo.
https://github.com/endless-sky/endless-sky/blob/ba9331f9e26e20d7d03353b93becdf9c003cf52f/source/PlayerInfo.cpp#L1833

It should become:
```c++
if(it->IsAtLocation(location) && it->CanOffer(*this) && !it->CanAccept(*this))
```

**Comment by warp-core** [comment]
The same change (`HasSpace` -> `CanAccept`) should also be made to line 1773 of `PlayerInfo.cpp`.

**Comment by rovermicrover** [comment]
@Amazinite & @warp-core It has been made so

**Comment by warp-core** [comment]
In order to actually get the blocked message to show (which, I think, is a big benefit of this), `Mission::BlockedMessage()` will also need to be udpated.

**Comment by rovermicrover** [comment]
I think we just need to add a condition to this early return to not return early it toAccept fails. Will take a look later when I get home.

https://github.com/rovermicrover/endless-sky/blob/to-accept-beta/source/Mission.cpp#L899

**Comment by rovermicrover** [comment]
I think that should do it.

https://github.com/endless-sky/endless-sky/pull/7835/commits/097366697bdf039e0238b2a08e4c193e10662c34

**Comment by rovermicrover** [comment]
Eee looking at it again maybe that is not the most elegant… hmmmmmm. Maybe the guard is no longer needed? Will need to take a longer look when I have more time…

**Comment by rovermicrover** [comment]
So I don’t even think we currently need that cargo and passenger space guard. It should never be in that state to begin with, and honestly the block message method shouldn’t care the reason, just that there is a block and to format the necessary data.

I think I have a solution that I can work on later tonight.

**Comment by rovermicrover** [comment]
I think the BlockedMessage method should now work even with the new to accept logic.

**Comment by rovermicrover** [comment]
Looking at this it’s gotten more complicated now. Will add some tests here.

**Comment by rovermicrover** [comment]
Test added to verify to accept block works.

**Comment by warp-core** [comment]
I've also tested it with an updated version of Cognizance 32 which can be made to work as intended with this.

File: source/Mission.cpp [hunk 1]
```diff
@@ -892,13 +905,19 @@ string Mission::BlockedMessage(const PlayerInfo &player)
 	if(flagship)
 		subs["<ship>"] = flagship->Name();
 
+	const auto &playerConditions = player.Conditions();
+	bool toAcceptPassed = toAccept.IsEmpty() || toAccept.Test(playerConditions);
+	subs["<conditions>"] = toAcceptPassed ? "meet" : "do not meet";
```
Comment (Amazinite) [inline source/Mission.cpp:910]: An empty condition set returns true when tested, so we can simplify things here.
```suggestion
	subs["<conditions>"] = toAccept.Test(playerConditions) ? "meet" : "do not meet";
```
What is the purpose of this substitution, though? How would you expect this to be used?

Comment (rovermicrover) [reply source/Mission.cpp:910]: “You need another 2 bunks, and do meet the requirement to have a jump drive install.”

“You have enough space, and you do meet the requirement to have a jump drive installed.”

**Pull request merged**